File: ossearch-crawler/src/main/java/edu/si/ossearch/reports/service/SD603ReportService.java
package edu.si.ossearch.reports.service;

import edu.si.ossearch.reports.dto.SD603ReportRequest;
import edu.si.ossearch.reports.dto.SD603ReportResponse;
import org.apache.solr.client.solrj.SolrServerException;

import java.io.ByteArrayInputStream;
import java.io.IOException;

/**
 * Service interface for generating SD603 Reports.
 * SD603 Reports search for specific words/phrases across selected collections
 * and return matching URLs organized by collection and search term.
 *
 * @author jbirkhimer
 */
public interface SD603ReportService {

    /**
     * Generate an SD603 report by searching for the specified terms across the given collections.
     *
     * @param request The report request containing collection IDs and search terms
     * @return The report response with matches organized by collection and term
     * @throws SolrServerException if there is an error communicating with Solr
     * @throws IOException if there is an I/O error
     */
    SD603ReportResponse generateReport(SD603ReportRequest request) throws SolrServerException, IOException;

    /**
     * Export an SD603 report to CSV format.
     *
     * @param report The report to export
     * @return A ByteArrayInputStream containing the CSV data
     * @throws IOException if there is an error generating the CSV
     */
    ByteArrayInputStream exportToCsv(SD603ReportResponse report) throws IOException;
}

File: ossearch-app-frontend-vue3/src/services/sd603report.service.js
import api from './api';

class SD603ReportService {
  /**
   * Generate SD603 report for specified collections and search terms
   * @param {number[]} collectionIds - Array of collection IDs to search
   * @param {string[]} searchTerms - Array of search terms/phrases
   * @returns {Promise} API response with report data
   */
  generateReport(collectionIds, searchTerms) {
    return api.post('/reports/sd603', {
      collectionIds: collectionIds,
      searchTerms: searchTerms
    });
  }

  /**
   * Export SD603 report as CSV file
   * @param {number[]} collectionIds - Array of collection IDs to search
   * @param {string[]} searchTerms - Array of search terms/phrases
   * @returns {Promise} API response with CSV blob
   */
  exportCsv(collectionIds, searchTerms) {
    return api.post('/reports/sd603/export', {
      collectionIds: collectionIds,
      searchTerms: searchTerms
    }, {
      responseType: 'blob'
    });
  }
}

export default new SD603ReportService();
